packages/explorer/src/app/(explorer)/[chainName]/worlds/[worldAddress]/observe/TransactionsTableContainer.tsx

```diff
@@ -0,0 +1,149 @@
+"use client";
+
+import { useParams } from "next/navigation";
+import {
+  AbiFunction,
+  BaseError,
+  DecodeFunctionDataReturnType,
+  Hex,
+  Log,
+  Transaction,
+  TransactionReceipt,
+  decodeFunctionData,
+  parseAbiItem,
+  parseEventLogs,
+} from "viem";
+import { useConfig, useWatchBlocks } from "wagmi";
+import { useStore } from "zustand";
+import React, { useMemo, useState } from "react";
+import { getTransaction, getTransactionReceipt, simulateContract } from "@wagmi/core";
+import { Write, store } from "../../../../../../observer/store";
+import { useChain } from "../../../../hooks/useChain";
+import { useWorldAbiQuery } from "../../../../queries/useWorldAbiQuery";
+import { TransactionsTableView } from "./TransactionsTableView";
+
+export type WatchedTransaction = {
+  hash?: Hex;
+  timestamp?: bigint;
+  transaction?: Transaction;
+  functionData?: DecodeFunctionDataReturnType;
+  receipt?: TransactionReceipt;
+  status: "pending" | "success" | "reverted";
+  write?: Write;
+  logs?: Log[];
+  error?: BaseError;
+};
+
+export function TransactionsTableContainer() {
+  const { id: chainId } = useChain();
+  const { worldAddress } = useParams();
+  const wagmiConfig = useConfig();
+  const { data: worldAbiData } = useWorldAbiQuery();
+  const abi = worldAbiData?.abi;
+  const [transactions, setTransactions] = useState<WatchedTransaction[]>([]);
+  const observerWrites = useStore(store, (state) => Object.values(state.writes));
+
+  const mergedTransactions = useMemo((): WatchedTransaction[] => {
+    const mergedMap = new Map<Hex | undefined, WatchedTransaction>();
+
+    for (const write of observerWrites) {
+      const parsedAbiItem = parseAbiItem(`function ${write.functionSignature}`) as AbiFunction;
+      const functionData = {
+        functionName: parsedAbiItem.name,
+        args: write.args,
+      };
+
+      mergedMap.set(write.hash, {
+        status: "pending",
+        functionData,
+        write,
+      });
+    }
+
+    for (const transaction of transactions) {
+      const existing = mergedMap.get(transaction.hash);
+      if (existing) {
+        mergedMap.set(transaction.hash, { ...transaction, write: existing.write });
+      } else {
+        mergedMap.set(transaction.hash, { ...transaction });
+      }
+    }
+
+    return Array.from(mergedMap.values()).reverse();
+  }, [transactions, observerWrites]);
+
+  async function handleTransaction(hash: Hex, timestamp: bigint) {
```

Comment:
hmm, feels weird that we're not writing to the transaction list until after all the things below resolve (getting tx, getting receipt, simulating result

wouldn't we want to render it right away, then enrich it with data as it comes in?

could we do this all inline with individual components rather than in a big handler like this? or does that impact the resulting order?

Reply:
good point, did it like that initially for simplicity reasons but definitely better to update state as soon as it's available.

As for individual components, I explored that approach but it didn't quite work due to how tanstack table works. It expects all the data to be passed to the `useReactTable` beforehand, and then provides utilities for rendering rows and cells. We could do without the tanstack table ofc but wanted to use it in case we need filtering, sorting and any other capabilities. Also, as you noted, there would be other considerations when it comes to sorting that would be need handling, mainly seems like handling incoming txs from observer would need more thought where tx hash is not immediately available.

Reply:
OOC do we need two components? seems like we could do one component and one hook? since most of the logic in `TransactionTableContainer` is now hook-like and we only wrap it in a component to render `TransactionTableView`